src/react/useMonaco.tsx

```diff
@@ -52,196 +48,227 @@ export interface UseMonacoOptions
+        const disposables = await onLoad(monaco);
+        if (disposables) {
+          onLoadDisposable = asDisposable(
+            Array.isArray(disposables) ? disposables : [disposables]
+          );
         }
-      } else if (theme) {
-        monaco.editor.setTheme(theme);
       }
+
+      // Save monaco to window and state.
+      window.monaco = monaco;
+      setMonaco(monaco);
+      setIsLoading(false);
+    }
+
+    initializeMonaco().catch((error) =>
+      console.error('An error occurred during initialization of Monaco:', error)
+    );
+
+    disposablesRef.current = [
+      pluginDisposable?.dispose,
+      onLoadDisposable?.dispose,
+      cancelable?.cancel,
+    ];
+  }, [monaco, languages, plugins]);
+
+  //
+  // Setup onThemeChange event handler
+  React.useEffect(() => {
+    if (!monaco) return;
+    if (!onThemeChange) return;
+    if (rFirstMount.current) return;
+
+    const disposable = monaco.editor.onDidChangeTheme((theme) => {
+      onThemeChange(theme, monaco);
+    });
+
+    return () => {
+      disposable?.dispose?.();
+    };
+  }, [monaco, onThemeChange, theme]);
+
+  //
+  // Setup theme and themes
+  React.useEffect(() => {
+    if (!monaco) return;
+    if (rFirstMount.current) return;
+
+    // Setup themes
+    if (!!themes) {
+      monaco.editor.defineThemes(themes);
+    }
+
+    // Set the current theme
+    if (!!theme) {
+      let themeToSet = typeof theme === 'function' ? theme() : theme;
+
+      if (typeof themeToSet === 'string' || !('then' in themeToSet)) {
+        monaco.editor.setTheme(themeToSet);
+      } else {
+        themeToSet.then(monaco.editor.setTheme);
+      }
+    }
+  }, [monaco, theme, themes]);
+
+  //
+  // A hook to run changes when monaco changes. (Maybe not needed?)
+  const useMonacoEffect = React.useCallback(
+    (cb: (monaco?: Monaco) => void | (() => void), deps: any[] = []) => {
+      return React.useEffect(() => monaco && cb(monaco), [monaco, ...deps]);
     },
-    [theme]
+    [monaco]
   );
 
+  // Cleanup disposables on unmount.
+  React.useEffect(() => {
+    rFirstMount.current = false;
+    return () => disposablesRef.current.forEach((fn) => fn && fn());
+  }, []);
+
   return {
-    monaco: monacoRef.current,
+    monaco,
     useMonacoEffect,
     defaultEditorOptions,
-    isLoading: Boolean(monacoRef.current),
+    isLoading,
   };
 };
 
-export function useRefWithEffect<T>(
-  initialValue: T
-): [
-  React.MutableRefObject<T | undefined>,
-  (effect: (obj: T) => void, deps: any[]) => void
-] {
-  const ref = React.useRef<T>(initialValue);
-  const useRefEffect = (effect: (obj: T) => void, deps: any[]) => {
-    React.useEffect(() => {
-      if (ref.current) {
-        return effect(ref.current);
-      }
-    }, [ref.current, ...deps]);
-  };
+// Helpers
 
-  return [ref, useRefEffect];
+function getPlugins(
+  plugins: UseMonacoOptions['plugins'],
+  languages: UseMonacoOptions['languages']
+) {
+  return [
+    ...plugins
```

Comment:
Even though ideally the plugins should specify their dependencies, I think we should pass the languages first then plugins